affinity/cpp-20/dXXX1r0.md

```diff
@@ -195,58 +266,147 @@ Below *(Listing 2)* is an example of executing a parallel task over 8 threads us
+* `sf` denotes a function object of type `SF`.
+
+| Nested Property Type | Nested Property Name | Requirements |
+|----------------------|----------------------|--------------|
+| bulk_execution_affinity_t::none_t | bulk_execution_affinity_t::none | A call to `e.bulk_execute(f, s, sf)` may or may not bind the created *execution agents* to the underlying *execution resources*. The affinity binding pattern may or may not be consistent across invocations of the executor's bulk execution function.  |
+| bulk_execution_affinity_t::scatter_t | bulk_execution_scatter_t::scatter | A call to `e.bulk_execute(f, s, sf)` must bind the created *execution agents* to the underlying *execution resources* such that they are distributed across the *execution resources* where each *execution agent* far from it's preceding and following *execution agents*. The affinity binding pattern must be consistent across invocations of the executor's bulk execution function. |
+| bulk_execution_affinity_t::compact_t | bulk_execution_compact_t::compact | A call to `e.bulk_execute(f, s, sf)` must bind the created *execution agents* to the underlying *execution resources* such that they are in sequence across the *execution resources* where each *execution agent* close to it's preceding and following *execution agents*. The affinity binding pattern must be consistent across invocations of the executor's bulk execution function. |
+| bulk_execution_affinity_t::balanced_t | bulk_execution_balanced_t::balanced | A call to `e.bulk_execute(f, s, sf)` must bind the created *execution agents* to the underlying *execution resources* such that they are in sequence and evenly spread across the *execution resources* where each *execution agent* is close to it's preceding and following *execution agents* and all *execution resources* are utilized. The affinity binding pattern must be consistent across invocations of the executor's bulk execution function. |
+
+> [*Note:* The requirements of the `bulk_execution_affinity_t` nested properties do not enforce a specific binding, simply that the binding follows the requirements set out above and that the pattern is consistent across invocations of the bulk execution functions. *--end note*]
+
+> [*Note:* If two *executors* `e1` and `e2` invoke a bulk execution function in order, where `execution::query(e1, execution::context) == query(e2, execution::context)` is `true` and `execution::query(e1, execution::bulk_execution_affinity) == query(e2, execution::bulk_execution_affinity)` is `false`, this will likely result in `e1` binding *execution agents* if necessary to achieve the requested affinity pattern and then `e2` rebinding to achieve the new affinity pattern. *--end note*]
+
+> [*Note:* The terms used for the `bulk_execution_affinity_t` nested properties are derived from the OpenMP properties [[33]][openmp-affinity] including the Intel specific balanced affinity binding [[[34]][intel-balanced-affinity] *--end note*]
+
+## Concurrency property
```

Comment:
Readers may wonder how this is relevant to the topic of affinity.  For example, the "explicitly manage memory by affinity region" example I proposed above would call for a different number, namely the number of affinity regions N, not the total available concurrency (>> N).

Reply:
Yeah, this property isn't specific to affinity, I just wanted to include it because it was previously available to the execution_context in the previous revision.

This is a good point. If we were to introduce an enumeration over different affinity domains, perhaps we could also have a property for querying the breadth of a particular affinity domain?

Reply:
@AerialMantis wrote:
> ... perhaps we could also have a property for querying the breadth of a particular affinity domain?

You would need something like that if you wanted to program explicitly to NUMA domains, as in the "allocate N chunks of memory with explicit affinity" example.  I suspect people will debate about whether that's a "high-level" interface.  It's a bit like calling `omp_get_num_threads` in an OpenMP parallel region instead of just writing a parallel loop in `#pragma omp parallel for` fashion.

Reply:
Yeah, I think you're right, it's more of a low-level concept. Though the more I think about it the more I think we might struggle to define the behaviour of these properties without tieing the wording to some flexible concept of an affinity domain. For example, the current properties assume a NUMA like architecture, so may not apply as well to a more heterogeneous architecture.

This is a bit orthogonal to this PR, but I've been thinking an interesting way to represent this could be via a combination of two properties. One for requiring an executor to be structured around a particular affinity domain, i.e. NUMA regions, CPUs, GPUs, cores, threads, etc, that would fail if that domain wasn't valid. And then another for querying or requiring a mask over the resources in the current affinity domain.

For example, you could require an executor to use a NUMA affinity domain, which could have a breadth of say 8 resources (which would be the NUMA regions), and then set the mask to only represent the first half. Or you could have an executor that has both a CPU and a GPU, and you could require it to use a GPU affinity domain, which would enforce that it executes using the GPU.